src/rust/engine/process_execution/src/nailgun/nailgun_pool.rs

```diff
@@ -1,258 +1,222 @@
-          );
-          if requested_server_fingerprint == process_fingerprint {
-            debug!("The fingerprint of the running nailgun {:?} matches the requested fingerprint {:?}. Connecting to existing server.",
-                   requested_server_fingerprint, process_fingerprint);
-            Ok(process_port)
-          } else {
-            // The running process doesn't coincide with the options we want.
-            if build_id_that_started_the_server == build_id_requesting_connection {
-              Err(format!(
-                "Trying to change the JVM options for a running nailgun server that was started this run, with name {}.\
-                  There is exactly one nailgun server per task, so it shouldn't be possible to change the options of a nailgun server mid-run.\
-                  This might be a problem with how we calculate the keys of nailgun servers (https://github.com/pantsbuild/pants/issues/8527).",
-                &name)
-              )
-            } else {
-              // Restart it.
-              // Since the stored server was started in a different pants run,
-              // no client will be running on that server.
-              debug!(
-                "The options for server process {} are different to the startup_options, \
-               and the original process was started in a different pants run.\n\
-               Startup Options: {:?}\n Process Cmd: {:?}",
-                &process_name, startup_options, process_fingerprint
-              );
-              debug!("Restarting the server...");
-              Self::start_new_nailgun(
-                &mut *processes,
-                name,
-                startup_options,
-                &workdir_path,
-                requested_server_fingerprint,
-                build_id_requesting_connection,
-              )
-            }
-          }
-        }
-        Ok(_) => {
-          // The process has exited with some exit code
-          debug!("The requested nailgun server was not running anymore. Restarting process...");
-          Self::start_new_nailgun(
-            &mut *processes,
-            name,
-            startup_options,
-            &workdir_path,
-            requested_server_fingerprint,
-            build_id_requesting_connection,
-          )
-        }
-        Err(e) => Err(e),
-      }
-    } else {
-      // We don't have a running nailgun registered in the map.
-      debug!(
-        "No nailgun server is running with name {}. Starting one...",
-        &name
-      );
-      Self::start_new_nailgun(
-        &mut *processes,
-        name,
+    // There wasn't a matching, valid, available process. We need to start one.
+    if processes.len() >= self.size {
+      // Find the oldest idle non-matching process and remove it.
+      let idx = Self::find_lru_idle(&mut *processes)?.ok_or_else(|| {
+        // NB: See the method docs: the pool assumes that it is running under a semaphore, so this
```

Comment:
Not really... my primary concern was really just that there are already two mutexes in here, and adding a third primitive would make things a little bit unwieldy. I think that if the consequence of forgetting to do this was a panic or silently incorrect behavior, then it would be necessary to include it here. But given that we can error instead, and we know how we are called in general, this seemed slightly easier.